### PR TITLE
Fix llama model sdpa attention forward function masking bug when output_attentions=True

### PR DESCRIPTION
**Pull request by Aladoro**
# What does this PR do?

Very simple fix to a nasty issue I have recently encountered. Due to its simplicity, I opened a PR directly without raising an issue first to avoid redundancy. Please, let me know if I should also raise an issue, and I'll do that right away.

## Description

When output_attentions is True, sdpa implementation's forward method calls the eager implementation's forward method. However, a None mask is still returned if sdpa's 'AttentionMaskConverter._ignore_causal_mask_sdpa' returns true (which occurs whenever the input is unmasked, as sdpa would defer the causal masking to the sdpa Pytorch implementation).
This inconsistency causes the model to run the eager implementation with no causal attention mask if the original input is unmasked (e.g., if a single input sequence is encoded or all encoded input sequences have the same length) and requires_attn=True. 

- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

Tagging @ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by Aladoro** [comment]
A minimal example of this erroneous behavior can be reproduced via:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model_name = "meta-llama/Meta-Llama-3-8B"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, 
                                             device_map='cuda', 
                                             torch_dtype=torch.bfloat16
                                             )

tokenizer.pad_token_id = tokenizer.eos_token_id

inputs = tokenizer(["Today is the day I went to the store and ..."],
                    return_tensors="pt").to('cuda')

expanded_batch_size = 1


outputs = model.generate(
    input_ids = inputs['input_ids'].expand(expanded_batch_size, -1),
    attention_mask = inputs['attention_mask'].expand(expanded_batch_size, -1),
    do_sample=False,
    max_new_tokens=5, 
    return_dict_in_generate=True,
    )


input_length = inputs.input_ids.shape[1]
sequences= outputs.sequences

for sequence in sequences:
    decoded_sequence = tokenizer.decode(sequence)
    print(decoded_sequence)

# separator
print('-'*20)


outputs = model.generate(
    input_ids = inputs['input_ids'].expand(expanded_batch_size, -1),
    attention_mask = inputs['attention_mask'].expand(expanded_batch_size, -1),
    do_sample=False,
    max_new_tokens=5, 
    return_dict_in_generate=True,
    output_attentions=True, # ?!
    )


input_length = inputs.input_ids.shape[1]
sequences= outputs.sequences

# garbage generated outputs since no masking is applied
for sequence in sequences:
    decoded_sequence = tokenizer.decode(sequence)
    print(decoded_sequence)
```

**Review by ArthurZucker — COMMENTED** [review]
Great catch. 
1. `causal_mask = AttentionMaskConverter._unmask_unattended(causal_mask, min_dtype)` line 1127 needs to be ignored as well.
2. we need to add your small example script as a test! 🤗

**Comment by ArthurZucker** [comment]
For 2. the test is already implemented, but I don't think it tests `output_attention=True`. It probably a matter of adding a `parametrized`. See this file here: https://github.com/huggingface/transformers/blob/main/tests/test_modeling_common.py#L3590

**Comment by Aladoro** [comment]
> Feel free to rebase it might be fixed on main / be flaky

Just did :)

**Comment by Aladoro** [comment]
@ArthurZucker Let me know if you think this fix is ready for merging, or if you'd like to add the tests to the same PR!

**Review by ArthurZucker — COMMENTED** [review]
Would be nice to just add the test in this PR 😉

**Comment by Aladoro** [comment]
> Would be nice to just add the test in this PR 😉

Alright - I made the addition of output_attentions=True to the sdpa equivalence test, as you suggested ;) (Black code re-formatting seems to have messed up the diff, but the changes are minimal...)

@ArthurZucker - Let me know if there are any outstanding issues or if there is something else missing before merging ^^

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -3753,177 +3753,194 @@ def get_mean_reldiff(failcase, x, ref, atol, rtol):
...
-                            if dummy_input.dtype in [torch.float32, torch.bfloat16, torch.float16]:
-                                dummy_input = dummy_input.to(torch_dtype)
+                        for output_attentions in [True, False]:
+                            if (
+                                not (
+                                    self.has_attentions
+                                    and "output_attentions" in inspect.signature(model_sdpa.forward).parameters
+                                )
+                                and output_attentions
+                            ):
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:3768]: can we split this in 2 lines? 
```suggestion
                            can_output_attn = "output_attentions" in inspect.signature(model_sdpa.forward).parameters
                            if (not (self.has_attentions and can_output_attn) and output_attentions):
```

**Review by ArthurZucker — COMMENTED** [review]
Ok, let's make sure you rebase as Gemma was updated a bit and commit with [run-slow] so that slow tests are run!

**Review by ArthurZucker — APPROVED** [review]
LGTM, fyi @fxmarty @gante and @ydshieh

**Comment by ArthurZucker** [comment]
(Merging once the CIs are all green!_

**Comment by Aladoro** [comment]
@ArthurZucker thanks for your suggestions! I also propagated the same changes to the new jetmoe model. All default checks are now passing ^^

**Comment by ArthurZucker** [comment]
THanks for the fix
